Unit test for the Encrypt class

enov opened the pull request:
The test data are provided by:
Cryptographic Algorithm Validation Program
National Institute of Standards and Technology

http://csrc.nist.gov/groups/STM/cavp/index.html

Used AES Multiblock Message Test (MMT) Sample Vectors
from http://csrc.nist.gov/groups/STM/cavp/documents/aes/aesmmt.zip
(the actual file used in that zipped file is CBCMMT128.rsp)

Also adds a class Kohana_EncryptTest_TestableEncrypt that wraps the
Encrypt class to mock out the actual mcrypt_create_iv calls.

However, it also changes Kohana_Encrypt internal API by adding
a protected function _create_iv to proxy mcrypt_create_iv
and to allow mocking and testing against KAT vectors.

Addresses #543.

Ready for merge. Kindly review thoroughly.

Thanks!

enov committed:
add protected function _create_iv to proxy mcrypt_create_iv

added a protected function `_create_iv`

This is to proxy the mcrypt_create_iv function
and to allow mocking and testing against KAT vectors.

enov committed:
Unit test for the Encrypt class

Adds a class Kohana_EncryptTest_TestableEncrypt that wraps the
Encrypt class to mock out the actual mcrypt_create_iv calls.

The test data are provided by
Cryptographic Algorithm Validation Program
National Institute of Standards and Technology

http://csrc.nist.gov/groups/STM/cavp/index.html

acoulton commented:
@enov this looks good and thorough to me - there might be an argument that it tests too much of the internal mcrypt behaviour but I think given the security implications it's probably better to do that than to mock out the mcrypt calls as well.

A few things I'd consider would really make this comprehensive
- Explicitly verifying that value > encode > decode === value - I know that the providers essentially do that by using known values, but IMHO it's a clearer demonstration of the behaviour of the class. Maybe you could combine the encode and decode tests?
- A test without the iv stubbed to verify that consecutive calls to encode the same value produce different results
- Tests for instance - that it returns singleton, that an instance created through the instance method has the expected configuration
- Tests for decode when the string is not valid base64, or is too short to contain a valid iv

enov commented:
Thanks for reviewing this @acoulton . I will rework on this soon to include the tests you described.

acoulton commented:
@enov no problem - they're a bit nerdy, you've done great work finding and implementing the NIST tests :)

enov committed:
More Encrypt tests

test_encode_decode:
Explicitly verifying that value > encode > decode === value

test_decode_invalid_data:
Tests for decode when the string is not valid base64, or is too short to
contain a valid IV

test_consecutive_encode_produce_different_results:
A test without the iv stubbed to verify that consecutive calls to encode
the same value produce different results

test_instance_throw_exception_when_no_key_provided
When an encryption key is not configured, the method `instance`
should throw a Kohana_Exception

test_instance_returns_singleton
Tests for method instance - that it returns singleton, that an instance
created through the instance method has the expected configuration

enov commented:
I don't know if we're affected by this?
https://bugs.php.net/bug.php?id=68238

enov commented:
http://php.net/manual/en/function.mcrypt-encrypt.php#refsect1-function.mcrypt-encrypt-changelog

## Changelog

**5.6.0**    Invalid key and iv sizes are no longer accepted. mcrypt_encrypt() will now throw a warning and return FALSE if the inputs are invalid. Previously keys and IVs were padded with '\0' bytes to the next valid size.

enov commented:
Reading @nikic comments, in https://bugs.php.net/bug.php?id=68238

> If that is the case, you should either modify it to explicitly emulate the old \0 byte padding behavior or - which I would strongly recommend - fix your encryption code to use correctly sized keys.

I suggest that we add the old \0 byte padding behavior ourselves in 3.3.

In 3.4, we could test for accepted key sizes (mcrypt_enc_get_supported_key_sizes?) and throw an exception at construction when not acceptable. Also, we probably need to provide/document some way to migrate old encrypted data to the new encryption behavior.

enov committed:
minor readability changes

enov committed:
Kohana_EncryptTest_TestableEncrypt to Kohana_EncryptTest_IvStubbed

Renamed Kohana_EncryptTest_TestableEncrypt to
Kohana_EncryptTest_IvStubbed so that I can add
another testable class

Also feed the known IV from the constructor now
instead of setting the property separately

enov committed:
Encryption key B/C compatible padding for PHP 5.6

Added a protected `_normalize_key` method to pad keys with
null bytes similar to PHP versions older than PHP 5.6 this
is called from the constructor when PHP version is above 5.6.0

Added Kohana_EncryptTest_KeyNormalized that always forces key
normalization, used in test_key_normalization.

Added test_key_normalization which tests our null byte padding
for encryption keys only against lower versions of PHP (< 5.6.0)
to see that it matches the internal key padding those PHP
versions already have.

enov commented:
@acoulton this is now ready for merge, with all your nerdy suggestions taken care of :)

Actually one of your suggestion revealed that PHP 5.6 does not pad null bytes to encryption keys anymore.

Kindly review.

acoulton commented on tests/kohana/EncryptTest.php:
```diff
@@ -0,0 +1,732 @@
+	 */
+	public function test_key_normalization($key, $iv, $mode, $cipher, $txt_plain)
+	{
+        if (version_compare(PHP_VERSION, '5.6.0', '>='))
```
@enov couple of whitespace issues here

enov replied:
those are fixed now in f3eea76

acoulton commented on tests/kohana/EncryptTest.php:
```diff
@@ -0,0 +1,732 @@
+		$expected_key = substr($config[$config_group]['key'], 0, $expected_key_size);
+
+		// assert
+		$this->assertSame($expected_key, $e->_key);
```
@enov shouldn't _key, _mode etc be protected properties? It looks like they're not actually declared on the class.

enov replied:
Yes, they are not declared on the class, they should have been originally declared as protected.

If I do so now, should I create the getters in `Kohana_EncryptTest_IvStubbed`? Or directly in the original `Kohana_Encrypt` class?

enov replied:
I guess it should go directly to the original `Kohana_Encrypt` class. But getters should be `public` and that would be too much to assume we're not breaking any existing code... Maybe we fix those in version 3.4?

acoulton replied:
hm, tricky. To test the values provided through `instance` you'll need to be checking on `Kohana_Encrypt` unless you can find a way to make `instance` create instances of the subclass.

But it doesn't feel right to expose getters for them - especially the key. They're deliberately internal to the class once it's constructed.

Sometimes for specific cases like this I assert the value of the protected property using reflection. It couples the test to the implementation of the class, and feels a bit dirty. But it limits the dirtiness to the test rather than exposing new public API to all consumers.

So eg:

``` php
$this->assertSameProtectedProperty($expected_key, $e, '_key');

protected function assertSameProtectedProperty($expect, $object, $name)
{
  $refl = new ReflectionClass($object);
  $property = $refl->getProperty($name);
  $property->setAccessible(TRUE);
  $this->assertSame($expect, $property->getValue($object));
}
```

enov replied:
@acoulton could you check now, I implemented the `assertSameProtectedProperty` and fixed the tests to use that.

enov replied:
Should I also test the `_iv_size`?

acoulton replied:
@enov great. I don't think you need to test `_iv_size` - it's calculated from mcrypt inside the class based on the cipher and method arguments. It's implementation detail that's covered by the fact your tests for encrypt/decrypt would fail if it broke.

You only need to spy into the class for the args provided to the constructor, everything else is covered by your behaviour tests already.

acoulton commented:
@enov this looks great, nice work! Should give us a lot more confidence. Just flagged a couple of minor issues.

enov committed:
tabs instead of whitespaces as per coding standards

enov committed:
Declare _key _mode _cipher _iv_size properties protected

Declared `_key`, `_mode`, `_cipher`, and `_iv_size` properties
that were originally created dynamically.

Added `assertSameProtectedProperty` as a helper method for the
tests, it uses reflection to access and test the values of those
properties
